activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -149,12 +149,14 @@ def grouping_queries(queries)
       end
 
       def convert_dot_notation_to_hash(attributes)
-        dot_notation = attributes.select do |k, v|
-          k.include?(".") && !v.is_a?(Hash)
+        dot_notation = attributes.each_with_object({}) do |(k, v), dn|
+          next if v.is_a?(Hash)
+          idx = k.rindex(".")
+          dn[k] = idx if idx
         end
 
-        dot_notation.each_key do |key|
-          table_name, column_name = key.split(".")
+        dot_notation.each do |key, idx|
+          table_name, column_name = key[0, idx], key[idx + 1, key.length]
```

Comment:
Can't we merge these two loops and avoid the temporary Hash?

Reply:
We can't mutate the original attributes hash while iterating over it.

Reply:
> if we're already mutating attributes that are passed to #build_from_hash maybe that's already fine?

Yeah that was my point. But also I don't think we need to mutate `attributes` at all, just iterate once over it and create the returned hash as we go.